File: src/pineland_sim/access.py
"""Persistent actor-owned access restrictions on locality corridors."""
from __future__ import annotations

from math import exp

from .entities import AccessRestriction, clamp
from .relations import organizations_hostile


def corridor_key(
    organization_id: str, first_locality_id: str, second_locality_id: str
) -> tuple[str, str, str]:
    if first_locality_id == second_locality_id:
        raise ValueError("corridor requires distinct localities")
    first, second = sorted((first_locality_id, second_locality_id))
    return organization_id, first, second


def build_access_restriction(
    world,
    organization_id: str,
    first_locality_id: str,
    second_locality_id: str,
    effort: float,
    time: float,
) -> AccessRestriction:
    if second_locality_id not in world.adjacency.get(first_locality_id, {}):
        raise ValueError("access restrictions can only be built on adjacent localities")
    key = corridor_key(organization_id, first_locality_id, second_locality_id)
    restriction = world.access_restrictions.get(key)
    if restriction is None:
        restriction = AccessRestriction(
            organization_id,
            key[1],
            key[2],
            0.0,
            float(time),
            float(time),
        )
        world.access_restrictions[key] = restriction
    rate = world.config.access_restriction.build_rate
    increment = clamp(rate * clamp(effort))
    restriction.level = clamp(
        1.0 - (1.0 - restriction.level) * (1.0 - increment)
    )
    restriction.cumulative_effort += max(0.0, float(effort))
    restriction.updated_at = float(time)
    return restriction


def decay_access_restrictions(world, time: float, elapsed_days: float) -> int:
    if elapsed_days <= 0:
        return 0
    decay_rate = world.config.access_restriction.decay_per_day
    factor = exp(-decay_rate * elapsed_days)
    removed = 0
    for key, restriction in list(world.access_restrictions.items()):
        restriction.level = clamp(restriction.level * factor)
        restriction.updated_at = float(time)
        if restriction.level <= 1e-9:
            del world.access_restrictions[key]
            removed += 1
    return removed


def edge_restriction_level(
    world,
    first_locality_id: str,
    second_locality_id: str,
    moving_organization_id: str | None = None,
) -> float:
    first, second = sorted((first_locality_id, second_locality_id))
    levels = []
    for (owner_id, edge_first, edge_second), restriction in world.access_restrictions.items():
        if edge_first != first or edge_second != second or restriction.level <= 0:
            continue
        if moving_organization_id is None:
            levels.append(restriction.level)
        elif owner_id == moving_organization_id:
            continue
        elif (
            owner_id in world.organizations
            and moving_organization_id in world.organizations
            and organizations_hostile(world, owner_id, moving_organization_id)
        ):
            levels.append(restriction.level)
    if not levels:
        return 0.0
    complement = 1.0
    for value in levels:
        complement *= 1.0 - clamp(value)
    return clamp(1.0 - complement)


def route_restriction_level(
    world, route: list[str] | tuple[str, ...], moving_organization_id: str | None = None
) -> float:
    return max(
        (
            edge_restriction_level(world, first, second, moving_organization_id)
            for first, second in zip(route, route[1:])
        ),
        default=0.0,
    )


def locality_access_pressure(world, locality_id: str) -> float:
    neighbors = world.adjacency.get(locality_id, {})
    if not neighbors:
        return 0.0
    return sum(
        edge_restriction_level(world, locality_id, neighbor, None)
        for neighbor in neighbors
    ) / len(neighbors)


def access_diagnostics(world) -> dict:
    return {
        "active_restrictions": len(world.access_restrictions),
        "mean_level": (
            sum(item.level for item in world.access_restrictions.values())
            / max(1, len(world.access_restrictions))
        ),
        "by_corridor": {
            "|".join(key): item.level
            for key, item in sorted(world.access_restrictions.items())
        },
    }

File: rust/pineland-model/src/access.rs
//! Persistent actor-owned access restrictions on locality corridors.
//!
//! The Python reference stores these rows in a sparse insertion-ordered map.
//! Native keeps the same logical key and update rules in parallel vectors so
//! the state can be checkpointed and gathered without object conversion.

use pineland_core::config::SimulationConfig;
use pineland_core::rng::python_exp;
use pineland_core::state::{clamp01, ParticleState};
use pineland_core::topology::StaticTopology;

pub fn organizations_hostile(particle: &ParticleState, first: usize, second: usize) -> bool {
    if first == second {
        return false;
    }
    particle
        .relations
        .organization_a
        .iter()
        .zip(&particle.relations.organization_b)
        .zip(&particle.relations.status)
        .any(|((a, b), status)| {
            ((*a as usize == first && *b as usize == second)
                || (*a as usize == second && *b as usize == first))
                && *status == 4
        })
}

pub fn edge_restriction_level(
    particle: &ParticleState,
    first: usize,
    second: usize,
    moving_organization: Option<usize>,
) -> f64 {
    let (first, second) = if first <= second {
        (first as u32, second as u32)
    } else {
        (second as u32, first as u32)
    };
    let mut complement = 1.0;
    let mut found = false;
    for index in 0..particle.access_restrictions.owner.len() {
        if particle.access_restrictions.first_locality[index] != first
            || particle.access_restrictions.second_locality[index] != second
            || particle.access_restrictions.level[index] <= 0.0
        {
            continue;
        }
        let owner = particle.access_restrictions.owner[index] as usize;
        let applies = match moving_organization {
            None => true,
            Some(moving) => owner != moving && organizations_hostile(particle, owner, moving),
        };
        if applies {
            found = true;
            complement *= 1.0 - clamp01(particle.access_restrictions.level[index]);
        }
    }
    if found {
        clamp01(1.0 - complement)
    } else {
        0.0
    }
}

pub fn locality_access_pressure(
    particle: &ParticleState,
    topology: &StaticTopology,
    locality: usize,
) -> f64 {
    let neighbors = topology
        .locality_edges
        .neighbors(locality)
        .collect::<Vec<_>>();
    if neighbors.is_empty() {
        return 0.0;
    }
    let mut total = 0.0;
    for (neighbor, _) in neighbors.iter().copied() {
        total += edge_restriction_level(particle, locality, neighbor as usize, None);
    }
    total / neighbors.len() as f64
}

pub fn route_restriction_level(
    particle: &ParticleState,
    route: &[usize],
    moving_organization: Option<usize>,
) -> f64 {
    route
        .windows(2)
        .map(|pair| edge_restriction_level(particle, pair[0], pair[1], moving_organization))
        .fold(0.0, f64::max)
}

pub fn build(
    particle: &mut ParticleState,
    owner: usize,
    first: usize,
    second: usize,
    effort: f64,
    time: f64,
    config: &SimulationConfig,
) -> f64 {
    let (first, second) = if first <= second {
        (first as u32, second as u32)
    } else {
        (second as u32, first as u32)
    };
    let index = particle
        .access_restrictions
        .find(owner, first as usize, second as usize)
        .unwrap_or_else(|| {
            let index = particle.access_restrictions.owner.len();
            particle.access_restrictions.owner.push(owner as u32);
            particle.access_restrictions.first_locality.push(first);
            particle.access_restrictions.second_locality.push(second);
            particle.access_restrictions.level.push(0.0);
            particle.access_restrictions.cumulative_effort.push(0.0);
            particle.access_restrictions.updated_at.push(time);
            index
        });
    let increment = clamp01(config.access_restriction.build_rate * clamp01(effort));
    let old = particle.access_restrictions.level[index];
    particle.access_restrictions.level[index] = clamp01(1.0 - (1.0 - old) * (1.0 - increment));
    particle.access_restrictions.cumulative_effort[index] += effort.max(0.0);
    particle.access_restrictions.updated_at[index] = time;
    particle.access_restrictions.level[index]
}

pub fn decay(
    particle: &mut ParticleState,
    time: f64,
    elapsed_days: f64,
    config: &SimulationConfig,
) -> usize {
    if elapsed_days <= 0.0 {
        return 0;
    }
    let factor = python_exp(-config.access_restriction.decay_per_day * elapsed_days);
    let mut removed = 0;
    let mut index = 0;
    while index < particle.access_restrictions.owner.len() {
        particle.access_restrictions.level[index] *= factor;
        particle.access_restrictions.updated_at[index] = time;
        if particle.access_restrictions.level[index] <= 1.0e-9 {
            particle.access_restrictions.owner.remove(index);
            particle.access_restrictions.first_locality.remove(index);
            particle.access_restrictions.second_locality.remove(index);
            particle.access_restrictions.level.remove(index);
            particle.access_restrictions.cumulative_effort.remove(index);
            particle.access_restrictions.updated_at.remove(index);
            removed += 1;
        } else {
            index += 1;
        }
    }
    removed
}
